src/transformers/trainer.py

```diff
@@ -2051,6 +2051,10 @@ def save_model(self, output_dir: Optional[str] = None):
         elif self.args.should_save:
             self._save(output_dir)
 
+        # Push to the Hub when `save_model` is called by the user.
+        if self.args.push_to_hub and not _internal_call:
```

Comment:
I don't fully understand this. Don't we also want to push to the Hub automatically during training (which is an internal call)?

Reply:
`push_to_hub` calls `save_model` which calls `push_to_hub` which calls `save_model` which calls `push_to_hub` which calls `save_model` which calls `push_to_hub` which calls `save_model` which calls `push_to_hub` which calls `save_model` which calls `push_to_hub` which calls `save_model` which calls `push_to_hub` which calls `save_model` which calls `push_to_hub` which calls `save_model` which calls `push_to_hub` which calls `save_model` ...

This internal argument is here to avoid that ;-)

Reply:
Thanks for explaining and for the PR! 

Unfortunately I'm still getting the behaviour in which the model is getting saved during training but it's not getting pushed. https://colab.research.google.com/drive/1GAXf3egH2GDbk7M0btdKWbLerBqLoJPi?usp=sharing. I have a commit saving at step 500 but no push

Reply:
Hi Omar! I just checked your colab, and ran it. You have to wait for a solid 15 minutes after the fact to see the weights on your repo as Colab is uploading at an excruciatingly slow speed (those pushes during training are synchronous to avoid slowing down training).

`trainer.push_in_progress` give you the job that is pushing, you can check its `stdout` attribute to see the progress it makes, its `is_done` attribute to see if it's finished or not and its `stderr` attribute to check if there was an error or not.

You can see on [this repo](https://huggingface.co/sgugger/debug-example2) that I eventually got my weights pushed with your code :-)

Reply:
Aha! Thanks @sgugger for the investigation. I was not aware about `trainer.push_in_progress`, this is something I'll definitively use next time. Thanks once again.